[ci] [python-package] scikit-learn 1.9+: 'base_estimator' argument removed from some classes

## Description

The CI job where we test against nightlies of dependencies started failing recently.

```text
____________________________ test_classifier_chain _____________________________

    def test_classifier_chain():
        n_outputs = 3
        X, y = make_multilabel_classification(n_samples=100, n_features=20, n_classes=n_outputs, random_state=0)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
        order = [2, 0, 1]
>       clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10), order=order, random_state=42)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       TypeError: ClassifierChain.__init__() got an unexpected keyword argument 'base_estimator'

tests/python_package_test/test_sklearn.py:456: TypeError
_____________________________ test_regressor_chain _____________________________

    def test_regressor_chain():
        bunch = load_linnerud(as_frame=True)  # returns a Bunch instance
        X, y = bunch["data"], bunch["target"]
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
        order = [2, 0, 1]
>       reg = RegressorChain(base_estimator=lgb.LGBMRegressor(n_estimators=10), order=order, random_state=42)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       TypeError: _BaseChain.__init__() got an unexpected keyword argument 'base_estimator'
```

([build link](https://github.com/lightgbm-org/LightGBM/actions/runs/24647183914/job/72062205338?pr=7228#step:4:205))

## Notes

PR that caused this: https://github.com/scikit-learn/scikit-learn/pull/33750

We'll need to adjust to this to fix CI, and will want to get a fix out in a new release soon (#7129).

--- a/.ci/test-r-package.sh
+++ b/.ci/test-r-package.sh
@@ -49,6 +49,7 @@ if [[ $OS_NAME == "linux" ]]; then
         --no-install-recommends \
         -y \
             devscripts \
+            libuv1-dev \
             r-base-core=${R_LINUX_VERSION} \
             r-base-dev=${R_LINUX_VERSION} \
             texinfo \

--- a/python-package/lightgbm/basic.py
+++ b/python-package/lightgbm/basic.py
@@ -3115,7 +3115,8 @@ class Dataset:
         # Check if the weight contains values other than one
         if weight is not None:
             if _is_pyarrow_array(weight):
-                if pa_compute.all(pa_compute.equal(weight, 1)).as_py():
+                # TODO: remove 'type: ignore[attr-defined]' when https://github.com/apache/arrow/issues/49831 is resolved.
+                if pa_compute.all(pa_compute.equal(weight, 1)).as_py():  # type: ignore[attr-defined]
                     weight = None
             elif np.all(weight == 1):
                 weight = None

--- a/tests/python_package_test/test_sklearn.py
+++ b/tests/python_package_test/test_sklearn.py
@@ -52,6 +52,7 @@ from .utils import (
 
 SKLEARN_MAJOR, SKLEARN_MINOR, *_ = _sklearn_version.split(".")
 SKLEARN_VERSION_GTE_1_6 = (int(SKLEARN_MAJOR), int(SKLEARN_MINOR)) >= (1, 6)
+SKLEARN_VERSION_GTE_1_7 = (int(SKLEARN_MAJOR), int(SKLEARN_MINOR)) >= (1, 7)
 
 decreasing_generator = itertools.count(0, -1)
 estimator_classes = (lgb.LGBMModel, lgb.LGBMClassifier, lgb.LGBMRegressor, lgb.LGBMRanker)
@@ -453,7 +454,15 @@ def test_classifier_chain():
     X, y = make_multilabel_classification(n_samples=100, n_features=20, n_classes=n_outputs, random_state=0)
     X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
     order = [2, 0, 1]
-    clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10), order=order, random_state=42)
+    # 'base_estimator' parameter was deprecated in scikit-learn 1.7 and removed in 1.9
+    #
+    #  * https://github.com/scikit-learn/scikit-learn/pull/30152
+    #  * https://github.com/scikit-learn/scikit-learn/pull/33750
+    #
+    if SKLEARN_VERSION_GTE_1_7:
+        clf = ClassifierChain(estimator=lgb.LGBMClassifier(n_estimators=10), order=order, random_state=42)
+    else:
+        clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10), order=order, random_state=42)
     clf.fit(X_train, y_train)
     score = clf.score(X_test, y_test)
     assert score >= 0.2

--- a/tests/python_package_test/test_sklearn.py
+++ b/tests/python_package_test/test_sklearn.py
@@ -479,7 +479,15 @@ def test_regressor_chain():
     X, y = bunch["data"], bunch["target"]
     X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
     order = [2, 0, 1]
-    reg = RegressorChain(base_estimator=lgb.LGBMRegressor(n_estimators=10), order=order, random_state=42)
+    # 'base_estimator' parameter was deprecated in scikit-learn 1.7 and removed in 1.9
+    #
+    #  * https://github.com/scikit-learn/scikit-learn/pull/30152
+    #  * https://github.com/scikit-learn/scikit-learn/pull/33750
+    #
+    if SKLEARN_VERSION_GTE_1_7:
+        reg = RegressorChain(estimator=lgb.LGBMRegressor(n_estimators=10), order=order, random_state=42)
+    else:
+        reg = RegressorChain(base_estimator=lgb.LGBMRegressor(n_estimators=10), order=order, random_state=42)
     reg.fit(X_train, y_train)
     y_pred = reg.predict(X_test)
     _, score, _ = mse(y_test, y_pred)

